packages/tools/scripts/rewrite-internal-redirects/index.ts

```diff
@@ -0,0 +1,307 @@
+
+import type { Node } from 'unist'
+import type { VFile } from 'vfile'
+
+const cwd = process.cwd()
+
+export const PRODUCT_DOMAIN_MAP = {
+  vault: 'vaultproject.io',
+  terraform: 'terraform.io',
+  consul: 'consul.io',
+  vagrant: 'vagrantup.com',
+  nomad: 'nomadproject.io',
+  waypoint: 'waypointproject.io',
+  cloud: 'cloud.hashicorp.com',
+  packer: 'packer.io',
+  boundary: 'boundaryproject.io',
+  sentinel: 'docs.hashicorp.com',
+} as const
+
+/**
+ * Determines if the given URL is an internal URL within the context of the provided product
+ * @param url URL to check
+ * @param product associated product, if any
+ * @returns
+ */
+export function isInternalUrl(
+  url: string,
+  product?: keyof typeof PRODUCT_DOMAIN_MAP
+) {
+  // relative paths are internal
+  if (url.startsWith('/')) return true
+
+  // Check the domain name of the URL if it's not relative. If it matches the domain for the supplied product, then it's not internal
+  try {
+    const { hostname } = new URL(url)
+    if (
+      !product &&
+      (Object.values(PRODUCT_DOMAIN_MAP) as string[]).some((domain) =>
+        hostname.endsWith(domain)
+      )
+    ) {
+      return true
+    }
+
+    if (product && hostname.endsWith(PRODUCT_DOMAIN_MAP[product])) return true
+  } catch {
+    // TODO: try and handle relative paths such as ./docker at some point
+  }
+
+  return false
+}
+
+/**
+ * Load and compile redirects from a redirects.js file in the cwd
+ */
+function loadRedirects() {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const rawRedirects = require(require.resolve(path.join(cwd, 'redirects.js')))
+
+  const redirects = rawRedirects.map((redirect) => {
+    const isExternalDestination = !redirect.destination.startsWith('/')
+    const doesDestinationHaveTokens = redirect.destination.includes('/:')
+
+    let destination: string | pathToRegexp.PathFunction = redirect.destination
+
+    // External URLs can't be passed to pathToRegexp directly, so we have to parse the URL
+    if (isExternalDestination) {
+      if (doesDestinationHaveTokens) {
+        destination = (params?: Record<string, any>) => {
+          const destUrl = new URL(redirect.destination)
+          const destCompile = pathToRegexp.compile(destUrl.pathname)
+
+          const newPath = destCompile(params)
+
+          destUrl.pathname = newPath
+
+          return destUrl.href
+        }
+      }
+    } else {
+      destination = pathToRegexp.compile(redirect.destination)
+    }
+
+    return {
+      source: pathToRegexp.match(redirect.source),
+      destination,
+      permanent: redirect.permanent,
+    }
+  })
+
+  return redirects
+}
+
+/**
+ * Checks for a matching redirect with the given URL and, if found,
+ * applies the matching redirect.
+ *
+ * @param url URL to check for redirects with and apply to
+ * @param redirects List of redirects which will be tested against the URL
+ * @returns The redirected URL
+ */
+export const checkAndApplyRedirect = (
```

Comment:
@thiskevinwang This function will not come into play for `terraform` at all, as we won't be looking at a redirect file.